LibELF+Kernel: Support initializing values of TLS data

itamar8910 opened the pull request:
We now support initializing TLS data with the values that exist in `PT_TLS` segments, i.e doing `__thread int myvar = 0xdeadbeef`.
Previously, TLS data was always zero-initialized.

Test program:
```
// tls-test.cpp:
#include <AK/Format.h>
#include <LibThread/Thread.h>
#include <MyLib/lib.h>

static int foo()
{
    outln("3: {:08x}", tls_var);
    return 0;
}

int main()
{
    outln("1: {:08x}", tls_var);
    tls_var = 0xbaddcafe;
    outln("2: {:08x}", tls_var);
    auto thread = LibThread::Thread::construct(foo);
    thread->start();
    return 0;
}

// lib.h:
#include <stdint.h>
extern __thread uint32_t tls_var;

// lib.cpp:
#include "lib.h"
__thread uint32_t tls_var = 0xdeadbeef;
```
// result:
```
1: deadbeef
2: baddcafe
3: deadbeef
```

cc @gunnarbeutner @ADKaster

gunnarbeutner commented:
That's one more building block for complete TLS support done. Very nice. :partying_face:

awesomekling commented on Userland/Libraries/LibELF/DynamicLoader.h:
```diff
@@ -152,7 +153,7 @@ class DynamicLoader : public RefCounted<DynamicLoader> {
     VirtualAddress m_dynamic_section_address;
 
     size_t m_tls_offset { 0 };
-    size_t m_tls_size { 0 };
+    size_t m_tls_size { 0 }; // TLS size of the current object
```
Instead of putting a comment here, where nobody will see it, let's instead call the variable `m_tls_size_of_current_object` :)

tomuta commented on Kernel/Syscalls/mmap.cpp:
```diff
@@ -595,14 +596,21 @@ KResultOr<FlatPtr> Process::sys$allocate_tls(size_t size)
+    {
+        Kernel::SmapDisabler disabler;
+        void* fault_at;
+        if (!Kernel::safe_memcpy((char*)m_master_tls_region.unsafe_ptr()->vaddr().as_ptr(), (char*)initial_data.ptr(), size, fault_at))
```
I think `copy_from_user` would be safer here (which would also validate the user mode pointer). That implicitly combines `SmapDisabler` with `safe_memcpy`.

itamar8910 replied:
Please correct me I'm confusing this, but I don't think `copy_from_user` would work here because `m_master_tls_region` is a user address, and `copy_from_user` [doesn't allow that](https://github.com/SerenityOS/serenity/blob/b0faf2287a016a1d1568313ab28a494c4b20398e/Kernel/StdLib.cpp#L195).

itamar8910 replied:
Thinking about it though there's not reason for `m_master_tls_region` to be user-accessible, so we could change that and then use `copy_from_user`

tomuta replied:
@itamar8910 I think you're right, I didn't realize the destination was also a user accessible pointer...

itamar8910 replied:
I think making `m_master_tls_region`  non user-accessible can be done in a different PR, so i'm marking this discussion as reolved :)

awesomekling commented:
This has a conflict now, sorry about that.

itamar8910 commented:
> This has a conflict now, sorry about that.

No worries, should be good now.

@gunnarbeutner can you review the fix in bf7570c?

gunnarbeutner commented:
With the PR applied I can't get to `dlopen()` any libraries:

![Screenshot from 2021-04-26 20-39-09](https://user-images.githubusercontent.com/388571/116134170-e710a180-a6cf-11eb-8d35-ae907e9f8c40.png)

itamar8910 commented:
> With the PR applied I can't get to `dlopen()` any libraries:

Yeah this reproduces for me as well. Thanks for checking this, i'll look into why this is.
@awesomekling please do not merge yet :)

itamar8910 commented:
> With the PR applied I can't get to `dlopen()` any libraries:

So that particular bug was a silly thing in f58cfce were i removed the `bool main_program` argument from `map_library`, but forgot to change one of its callers, which then called the `map_library(const String& filename, int fd)` overload with `false` for the fd :)

To support TLS data in `dlopen`ed libraries, I changed the TLS offset calculation in 80b9e09. After this change we don't need to change the TLS offsets of already loaded libraries if we `dlopen` a library that has TLS.

Also please note that we currently don't support initializing the TLS-data of `dlopen`ed libraries (So doing `__thread int i = 7` like you did in your example won't work).

Test program for `dlopen`ing a library that has TLS:
```
// lib.cpp
__thread int i = 0;

extern "C"
{
int get_i();
void set_i(int);

int get_i(){
    return i;
}

void set_i(int val){
    i = val;
}

}

// main.cpp
#include <stdio.h>
#include <LibDl/dlfcn.h>

int main()
{
    void* lib  = dlopen("lib.so", RTLD_NOW | RTLD_GLOBAL);
    printf("lib: %p\n", lib);
    if(!lib){
        printf("%s\n", dlerror());
        return 1;
    }
    using GetIFunc = int(*)();
    using SetIFunc = void(*)(int);
    auto get_i = (GetIFunc)dlsym(lib, "get_i");
    auto set_i = (SetIFunc)dlsym(lib, "set_i");
    printf("get_i: %p\n", get_i);
    printf("get_i(): %d\n", get_i());
    set_i(1);
    printf("get_i(): %d\n", get_i());
    return 0;
}

// output:
lib: 0x00a90218
get_i: 0x70ff14c0
get_i(): 0
get_i(): 1
```

gunnarbeutner commented:
Appears to be working for me. :+1: 

* Test program which creates a thread and reads/writes a `__thread` value in a `dlopen()`d shared library via some `dlsym()`d getter/setter functions.
* Python 3: `import json` and `json.dumps` works.
